### PR TITLE
threads: restrict number of golang threads if not set

### PR DESCRIPTION
**Pull request by grahamwhaley**
If the max number of golang threads (GOMAXPROCS) is not set in
the environment, then set it ourselves to a small set, but
only if the host has more cpus than that max.
This avoids us taking the default (the number of host cores)
on large-core systems, which can end up with us having a lot of
idle threads generated, consuming host PID space, and thus
artificially restricting how many containers we can run
in parallel.

Fixes: #114

Signed-off-by: Graham Whaley <graham.whaley@intel.com>

### CODE REVIEW

**Comment by grahamwhaley** [comment]
For both this and its equiv over in the proxy, I ran the metrics density and footprint tests to check if there was visible (hopefully positive) impact.
On the 88 core machine there was a mild (~2%) drop in KSM density and boot speed. That is probably near the 'noise' threshold of the tests, but at least it is in the right direction.

File: main.go [hunk 1]
```diff
@@ -80,6 +81,21 @@ func realMain() {
...
+	// which un-necessarily consume host PID space (and thus set an
+	// artificial max limit on the number of concurrent containers we can
+	// run)
+	if os.Getenv("GOMAXPROCS") == "" {
+		// We choose 6 as we want a couple of threads for the runtime (gc etc.)
+		// and couple of threads for our parallel user code, such as the copy
+		// code in shim.go
+		const maxthreads = 6
+		if runtime.NumCPU() > maxthreads {
+			runtime.GOMAXPROCS(6)
```
Comment (jodh-intel) [inline main.go]: `runtime.GOMAXPROCS(maxthreads)` maybe? :)

Same comment as for https://github.com/kata-containers/proxy/pull/118 - might be worth moving the `const` to the top for better visibility.

Comment (grahamwhaley) [reply main.go]: s/6/maxthreads/ indeed (oops ;-)).
The const - yeah, was not sure if better at the top or here.
I may also move the code to the actual `main()`, to fix the complexity CI niggle. It's not really clear to me why we have `realMain()` ;-)

**Comment by jodh-intel** [comment]
lgtm

**Comment by jodh-intel** [comment]
lgtm

**Comment by jodh-intel** [comment]
Keep up github! :) Let's make it a hat-trick...

lgtm!

**Comment by jodh-intel** [comment]
Hmm, all the metrics tests failed with the message "Failed to load JSON"... ?

**Comment by grahamwhaley** [comment]
Hmm, oh. I'll go look. I made a change yesterday, but don't think that would have created that error....

**Comment by grahamwhaley** [comment]
initrd CI failed with what looked like a timeout, so I've nudged a rebuild.
Metrics I know is a bit up/down right now, so pls ignore that. And I think the -0.91% codecov drop is probably unavoidable (if you look at the small code changes).

**Comment by grahamwhaley** [comment]
Slow week for PR reviews this week ... nudge nudge nudge....

**Comment by sboeuf** [comment]
metrics is broken right now, let's merge this anyway.

**Pull request merged**